### PR TITLE
Cypress tests refactoring: PO model, safe app tests

### LINKED ISSUE #2521: Refactoring of exiting Cypress tests (P3)
Refactoring of existing Cypress tests: PO model, safe apps tests, finish smoke tests, general refactoring

### PR DESCRIPTION
**Pull request by mike10ca**
## What it solves

- Refactoring of existing Cypress tests

Resolves #2521 

## How this PR fixes it

- PO model, safe apps tests, finish smoke tests, general refactoring

### CODE REVIEW

File: .github/workflows/e2e.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ concurrency:
 
 jobs:
   e2e:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
```
Comment (katspaugh) [inline .github/workflows/e2e.yml:12]: Why is this necessary? Latest is too new for Cypress? Or too old?

Comment (mike10ca) [reply .github/workflows/e2e.yml:12]: Our Cypress tests started to fail due to using latest runner image which conflicts with Cypress version. -latest tag uses image which might not be that stable as previous one, or be in pre-release mode.

File: cypress/e2e/pages/create_tx.pages.js [hunk 2]
```diff
@@ -149,3 +152,40 @@ export function verifyQueueLabel() {
 export function verifyTransactionSummary(sendValue) {
   cy.contains(TransactionSummary + `${sendValue} ${constants.tokenAbbreviation.gor}`).should('exist')
 }
+
+export function verifyDateExists(date) {
+  cy.contains('div', date).should('exist')
+}
+
+export function verifyImageAlttxt(index, text) {
```
Comment (iamacook) [inline cypress/e2e/pages/create_tx.pages.js:160]:
```suggestion
export function verifyImageAltTxt(index, text) {
```

File: cypress/e2e/pages/create_tx.pages.js [hunk 3]
```diff
@@ -149,3 +152,40 @@ export function verifyQueueLabel() {
...
+    .within(() => {
+      actions()
+    })
+}
+
+export function lickOnExpandAllBtn() {
+  cy.contains(expandallbtnStr).click()
+}
+
+export function lickOnCollapseAllBtn() {
```
Comment (iamacook) [inline cypress/e2e/pages/create_tx.pages.js:189]:
```suggestion
export function clickOnCollapseAllBtn() {
```

File: cypress/e2e/pages/create_tx.pages.js [hunk 4]
```diff
@@ -26,6 +27,8 @@ const editBtnStr = 'Edit'
 const executionParamsStr = 'Execution parameters'
 const noLaterStr = 'No, later'
 const signBtnStr = 'Sign'
+const expandallbtnStr = 'Expand all'
```
Comment (iamacook) [inline cypress/e2e/pages/create_tx.pages.js:30]:
```suggestion
const expandAllBtnStr = 'Expand all'
```

File: cypress/e2e/pages/safeapps.pages.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
+import * as constants from '../../support/constants'
+
+const searchappInput = 'input[id="search-by-name"]'
```
Comment (iamacook) [inline cypress/e2e/pages/safeapps.pages.js:3]:
```suggestion
const searchAppInput = 'input[id="search-by-name"]'
```

File: cypress/e2e/pages/safeapps.pages.js [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+const addBtnStr = /add/i
+const noAppsStr = /no Safe Apps found/i
+const bookmarkedAppsStr = /bookmarked Apps/i
+const customAppsStr = /my custom Apps/i
+const addCustomAppBtnStr = /add custom Safe App/i
+const openSafeAppBtnStr = /open Safe App/i
+const disclaimerTtle = /disclaimer/i
+const continueBtnStr = /continue/i
+const cameraCheckBoxStr = /camera/i
+const microfoneCheckBoxStr = /microphone/i
```
Comment (iamacook) [inline cypress/e2e/pages/safeapps.pages.js:16]:
```suggestion
const microphoneCheckBoxStr = /microphone/i
```

File: cypress/e2e/pages/safeapps.pages.js [hunk 7]
```diff
@@ -0,0 +1,193 @@
+import * as constants from '../../support/constants'
+
+const searchappInput = 'input[id="search-by-name"]'
+const appUrlInput = 'input[name="appUrl"]'
+const closePreviewWindowBtn = 'button[aria-label*="Close"][aria-label*="preview"]'
+
+const addBtnStr = /add/i
```
Comment (iamacook) [inline cypress/e2e/pages/safeapps.pages.js:7]: Nit: can we be a bit more specific that this is within the modal as there is also another add button in the list. (The same for the function which references this string.)

File: cypress/e2e/pages/safeapps.pages.js [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+const allowAllPermissions = /allow all/i
+
+const appNotSupportedMsg = "The app doesn't support Safe App functionality"
+
+export const pinWalletConnectStr = /pin walletconnect/i
+export const transactionBuilderStr = /pin transaction builder/i
+export const logoWalletConnect = /logo.*walletconnect/i
+export const walletConnectHeadlinePreview = /walletconnect/i
+export const availableNetworksPreview = /available networks/i
+export const connecttextPreview = 'Connect your Safe to any dApp that supports WalletConnect'
```
Comment (iamacook) [inline cypress/e2e/pages/safeapps.pages.js:30]:
```suggestion
export const connectTextPreview = 'Connect your Safe to any dApp that supports WalletConnect'
```

File: cypress/e2e/pages/safeapps.pages.js [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+export const localStorageItem =
+  '{"https://safe-test-app.com":[{"feature":"camera","status":"granted"},{"feature":"microphone","status":"denied"}]}'
+export const gridItem = 'main .MuiPaper-root > .MuiGrid-item'
+export const linkNames = {
+  logo: /logo/i,
+}
+
+export const permissionCheckboxes = {
+  camera: 'input[name="camera"]',
+  addressbook: 'input[name="requestAddressBook"]',
```
Comment (iamacook) [inline cypress/e2e/pages/safeapps.pages.js:40]:
```suggestion
  addressBook: 'input[name="requestAddressBook"]',
```

File: cypress/e2e/pages/safeapps.pages.js [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+  microphone: 'input[name="microphone"]',
+  geolocation: 'input[name="geolocation"]',
+  fullscreen: 'input[name="fullscreen"]',
+}
+
+export const permissionCheckboxNames = {
+  camera: 'Camera',
+  addressbook: 'Address Book',
```
Comment (iamacook) [inline cypress/e2e/pages/safeapps.pages.js:48]:
```suggestion
  addressBook: 'Address Book',
```

File: cypress/e2e/pages/safeapps.pages.js [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+const appUrlInput = 'input[name="appUrl"]'
+const closePreviewWindowBtn = 'button[aria-label*="Close"][aria-label*="preview"]'
+
+const addBtnStr = /add/i
+const noAppsStr = /no Safe Apps found/i
+const bookmarkedAppsStr = /bookmarked Apps/i
+const customAppsStr = /my custom Apps/i
+const addCustomAppBtnStr = /add custom Safe App/i
+const openSafeAppBtnStr = /open Safe App/i
+const disclaimerTtle = /disclaimer/i
```
Comment (iamacook) [inline cypress/e2e/pages/safeapps.pages.js:13]:
```suggestion
const disclaimerTitle = /disclaimer/i
```

File: cypress/e2e/pages/safeapps.pages.js [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+
+export function clickOnContinueBtn() {
+  return cy.findByRole('button', { name: continueBtnStr }).click()
+}
+
+export function verifyCameraCheckBoxExists() {
+  cy.findByRole('checkbox', { name: cameraCheckBoxStr }).should('exist')
+}
+
+export function verifyMicrofoneCheckBoxExists() {
```
Comment (iamacook) [inline cypress/e2e/pages/safeapps.pages.js:143]:
```suggestion
export function verifyMicrophoneCheckBoxExists() {
```

File: cypress/e2e/safe-apps/permissions_settings.cy.js [hunk 13]
```diff
@@ -1,36 +1,41 @@
...
 
 describe('The Safe Apps permissions settings section', () => {
   before(() => {
+    cy.clearLocalStorage()
     cy.on('window:before:load', (window) => {
       window.localStorage.setItem(
         constants.BROWSER_PERMISSIONS_KEY,
         JSON.stringify({
-          'https://app1.com': [
+          app1: [
```
Comment (iamacook) [inline cypress/e2e/safe-apps/permissions_settings.cy.js:16]: Let's keep the URL keys here to be consistent. You can use the variables as object keys to keep things concise:
```suggestion
          [app1]: [
```

Comment (iamacook) [reply cypress/e2e/safe-apps/permissions_settings.cy.js:16]: I won't mark the other changes here.

File: cypress/e2e/smoke/pending_actions.cy.js [hunk 14]
```diff
@@ -1,13 +1,14 @@
 import * as constants from '../../support/constants'
 import * as safe from '../pages/load_safe.pages'
-import * as main from '../pages/main.page'
 
 describe('Pending actions', () => {
   before(() => {
     cy.visit(constants.welcomeUrl)
-    main.acceptCookies()
+    // main.acceptCookies()
```
Comment (iamacook) [inline cypress/e2e/smoke/pending_actions.cy.js:7]: I assume this is commented out as the other tests are now skipped?

File: cypress/e2e/smoke/pending_actions.cy.js [hunk 15]
```diff
@@ -1,13 +1,14 @@
...
   })
 
+  //TODO: Discuss test logic
```
Comment (iamacook) [inline cypress/e2e/smoke/pending_actions.cy.js:10]: Need we skip the tests for now?

File: cypress/e2e/smoke/tx_history.cy.js [hunk 16]
```diff
@@ -1,25 +1,37 @@
 import * as constants from '../../support/constants'
+import * as main from '../pages/main.page'
+import * as createtx from '../pages/create_tx.pages'
```
Comment (iamacook) [inline cypress/e2e/smoke/tx_history.cy.js:3]:
```suggestion
import * as createTx from '../pages/create_tx.pages'
```

File: cypress/support/constants.js [hunk 17]
```diff
@@ -17,16 +17,26 @@ export const goerlyE2EWallet = /E2E Wallet @ G(ö|oe)rli/
...
 export const homeUrl = '/home?safe='
 export const welcomeUrl = '/welcome'
 export const chainMaticUrl = '/welcome?chain=matic'
+export const appsUrl = '/apps'
+export const requestPermissionsUrl = '/request-permissions'
+export const getPermissionsUrl = '/get-permissions'
+export const appSettingsUrl = '/settings/safe-apps'
+export const invalidappUrl = 'https://my-invalid-custom-app.com/manifest.json'
+export const validappUrlJson = 'https://my-valid-custom-app.com/manifest.json'
+export const validappUrl = 'https://my-valid-custom-app.com'
```
Comment (iamacook) [inline cypress/support/constants.js:36]:
```suggestion
export const invalidAppUrl = 'https://my-invalid-custom-app.com/manifest.json'
export const validAppUrlJson = 'https://my-valid-custom-app.com/manifest.json'
export const validAppUrl = 'https://my-valid-custom-app.com'
```

File: cypress/support/constants.js [hunk 18]
```diff
@@ -17,16 +17,26 @@ export const goerlyE2EWallet = /E2E Wallet @ G(ö|oe)rli/
...
+
 export const proposeEndPoint = '/**/propose'
+export const appsEndPoint = '/**/safe-apps'
```
Comment (iamacook) [inline cypress/support/constants.js:39]:
```suggestion
export const proposeEndpoint = '/**/propose'
export const appsEndpoint = '/**/safe-apps'
```

File: cypress/support/constants.js [hunk 19]
```diff
@@ -47,3 +57,32 @@ export const networks = {
 export const tokenAbbreviation = {
   gor: 'GOR',
 }
+
+export const appNames = {
+  walletconnect: 'walletconnect',
+  customcontract: 'compose custom contract',
+  noresults: 'atextwithoutresults',
```
Comment (iamacook) [inline cypress/support/constants.js:64]:
```suggestion
  walletConnect: 'walletconnect',
  customContract: 'compose custom contract',
  noResults: 'atextwithoutresults',
```

File: cypress/support/constants.js [hunk 20]
```diff
@@ -47,3 +57,32 @@ export const networks = {
...
+  received: 'Receive',
+  sent: 'Send',
+  deposit: 'deposit',
+  approve: 'Approve',
+  success: 'Success',
+  interaction: 'Contract interaction',
+}
+
+export const tokenNames = {
+  wrapped_ether: 'Wrapped Ether',
```
Comment (iamacook) [inline cypress/support/constants.js:87]:
```suggestion
  wrappedEther: 'Wrapped Ether',
```

**Review by iamacook — APPROVED** [review]

**Review by katspaugh — APPROVED** [review]

File: .github/workflows/e2e.yml (hunk 1 above)
Comment (mike10ca) [reply .github/workflows/e2e.yml:12]: By specifying a version, we ensure that there will be some control over it. When this version is planned to become unavailable, we will hopefully receive notification that will give us enough time for updating to newer version of Cypress/resolve any conflicts instead of reacting to this when it was unexpected.
